libraries/SoftwareSerial/SoftwareSerial.cpp

```diff
@@ -0,0 +1,445 @@
+    if(firstStartBit)
+    {
+      tunedDelay(initRxCenteringDelay);
+      firstStartBit = false;
+    }
+    else
+    {
+      tunedDelay(rxCenteringDelay);
+    }
+
+    for (uint8_t i=8; i > 0; --i)
+    {
+      tunedDelay(rxIntraBitDelay);
+      d >>= 1;
+      if (digitalRead(rxPin))  //rx_pin_read())
+        d |= 0x80;
+    }
+    
+    /**
+    Serial.println("recv");
+    Serial.print("int: ");
+    Serial.println(d);
+    Serial.print("char: ");
+    Serial.println((char)d);
+    Serial.print("Binary : ");
+    Serial.println(d, BIN);
+    **/
+    
+    //if (_inverse_logic)
+    //  d = ~d;
+
+    uint8_t next = (_receive_buffer_tail + 1) % _SS_MAX_RX_BUFF;
+    if (next != _receive_buffer_head)
+    {
+      // save new data in buffer: tail points to where byte goes
+      _receive_buffer[_receive_buffer_tail] = d; // save new byte
+      _receive_buffer_tail = next;
+    } 
+    else 
+    {
+      //DebugPulse(_DEBUG_PIN1, 1);
+      bufferOverflow = true;
+    }
+
+    // skip the stop bit
+    tunedDelay(bitDelay * 3 / 4);
+    //DebugPulse(_DEBUG_PIN1, 1);
+  }
+  interrupts();
+}
+
+uint32_t SoftwareSerial::rx_pin_read()
+{
+  return digitalRead(rxPin);
+}
+
+//
+// Interrupt handling
+//
+
+/* static */
+inline void SoftwareSerial::handle_interrupt()
+{
+  if (active_object)
+  {
+    active_object->recv();
+  }
+}
+
+//
+// Constructor
+//
+SoftwareSerial::SoftwareSerial(uint32_t receivePin, uint32_t transmitPin, bool inverse_logic /* = false */) : 
+  _rx_delay_centering(0),
+  _rx_delay_intrabit(0),
+  _rx_delay_stopbit(0),
+  _tx_delay(0),
+  _buffer_overflow(false),
+  _inverse_logic(inverse_logic)
+{
+  setTX(transmitPin);
+  txPin = transmitPin;
+  setRX(receivePin);
+  rxPin = receivePin;
+}
+
+//
+// Destructor
+//
+SoftwareSerial::~SoftwareSerial()
+{
+  end();
+}
+
+void SoftwareSerial::setTX(uint8_t tx)
+{
+  // First write, then set output. If we do this the other way around,
+  // the pin would be output low for a short while before switching to
+  // output hihg. Now, it is input with pullup for a short while, which
+  // is fine. With inverse logic, either order is fine.
+  digitalWrite(tx, _inverse_logic ? LOW : HIGH);
+  pinMode(tx, OUTPUT);
+}
+
+void SoftwareSerial::setRX(uint8_t rx)
+{
+  pinMode(rx, INPUT);
+  if (!_inverse_logic)
+    pinMode(rx, INPUT_PULLUP);
+    //digitalWrite(rx, HIGH);  // pullup for normal logic!
+  _receivePin = rx;
+  //_receiveBitMask = digitalPinToBitMask(rx);
+  //uint32_t port = digitalPinToPort(rx);
+  //_receivePortRegister = portInputRegister(port);
+}
+
+uint16_t SoftwareSerial::subtract_cap(uint16_t num, uint16_t sub) {
+  if (num > sub)
+    return num - sub;
+  else
+    return 1;
+}
+
+//
+// Public methods
+//
+
+void SoftwareSerial::begin(long speed)
+{
+  _rx_delay_centering = _rx_delay_intrabit = _rx_delay_stopbit = _tx_delay = 0;
+
+  //pre-calculate delays
+  bitDelay = (1000000/speed + 0.5);
```

Comment:
the delays need to be calculated after the user defines the baud rate they are using